### ISSUE TITLE
add functions for volumetric workflow

### ISSUE THREAD

**Issue by llevitis** [ReinderVosDeWael/BrainSpace#3]
- [ ] Function to map 4D volume data to 2D connectivity matrix
- [ ] Function to map 2D volume to 3D label image
- [ ] Update documentation with tutorial for Python volume viewer

**Comment by PeerHerholz** [comment]
thx for the roadmap @llevitis! anything you want to tackle? I would be okay with any of the tasks. @ReinderVosDeWael where exactly should these functions go within [here](https://github.com/ReinderVosDeWael/BrainSpace/tree/master/brainspace/gradient)?

**Comment by ReinderVosDeWael** [comment]
I'd place volume/label mapping functions in the utils directory https://github.com/ReinderVosDeWael/BrainSpace/tree/master/brainspace/utils . Although will ping Oualid to check with him.

**Comment by ReinderVosDeWael** [comment]
Yup utils directory is good.

**Comment by vinferrer** [comment]
If you want I can take care of the 4D to 2D connectivity matrix function, should be an easy one

**Comment by vinferrer** [comment]
or  i can do something else

**Comment by PeerHerholz** [comment]
![giphy](https://user-images.githubusercontent.com/20129524/78784586-966ac100-7973-11ea-8c47-5dddeb27866e.gif)

any updates of who is doing what?

**Comment by vinferrer** [comment]
yeah I was waiting for a response, but basically i already have Function to map 4D volume data to 2D connectivity matrix based on @llevitis code, I can upload it as a PR

**Comment by vinferrer** [comment]
https://github.com/vinferrer/BrainSpace/blob/vol2CM/brainspace/utils/volflow.py

**Comment by llevitis** [comment]
@PeerHerholz, I will make a commit to the branch created by @vinferrer to have an option of working with multiple timeseries. I played around yesterday with using Jinja and a template html file to create one html file that displays each interactive image created by Nilearn (each of which is saved to another html file). 

![image](https://user-images.githubusercontent.com/17363738/78797131-4b59a980-7985-11ea-8fa6-bd83facf866a.png)

**Comment by vinferrer** [comment]
Don't worry @llevitis  I can do the commit and then push it as a PR

**Comment by PeerHerholz** [comment]
Sorry for the late reply and the absence. I will now start working on the matrix to volume mapping, following what we have discussed in the other PRs and mattermost.

**Comment by ReinderVosDeWael** [comment]
Finally got around to running @llevitis code so far and it's working well for me and looks great :D. I think once @PeerHerholz is done we can confirm the PR.

**Comment by PeerHerholz** [comment]
Aaaaaaallrighty folks, after some while (super sorry again), I got to do my part. I have a running version of the function that maps computed `gradients` from `GradientMaps` to either `4D` or `3D` `volumes`. I'm sure it can be improved and we would need to talk about some of the specifics, but in general it runs. However, before I push my changes, I think we should check and merge #8 in order to avoid potential merge conflicts (sorry, I'm just being lazy).

**Comment by ReinderVosDeWael** [comment]
Merge #8 into my fork or into this pull request? If the former, is everyone OK with merging #8?

**Comment by vinferrer** [comment]
I think #8  isn't completely reviewed, please do not merge. @PeerHerholz and @llevitis have to approve it yet

**Comment by PeerHerholz** [comment]
On it!

**Comment by llevitis** [comment]
@PeerHerholz I am very very late to getting back to all of this. have you pushed your function yet?

**Comment by PeerHerholz** [comment]
Hi gang (@vinferrer, @llevitis, @ReinderVosDeWael),

while working with @ReinderVosDeWael on the `grad2fmrivols` function and some applications, we stumbled upon a few limitations of the current workflow that should be addressed:

-  `atlas`  should be more flexible to also allow custom parcellations, as the current version loves `nilearn` maybe a bit too much
- `voxel` based fun is not possible, for that we simply need to integrate [NiftiMasker](http://nilearn.github.io/modules/generated/nilearn.input_data.NiftiMasker.html) (tried it with ~4500 voxels and it worked fine)

Please add your input here!

**Comment by PeerHerholz** [comment]
Add on: while playing a bit around, I think we need to disentangle `atlas` and `labels` in order to allow more flexibility. 
I would propose the following:

- `atlas`: either a `str`/`path` to the `.nii` image or a loaded `nii`
- `labels`: a `np.array` with corresponding label names as `str`

With that we could also support `voxel-wise` stuff as `atlas` (we would need to change the name of the argument) could also be a `mask` and thus `labels` is optional and only needed if the other argument is an `atlas`. WDYT?

**Comment by ReinderVosDeWael** [comment]
I think the most important thing would be that we could simply provide a nifti image containing a parcellation. This would allow the user to export gradients to any atlas they have in a way that's probably intuitive to most neuroscientists.  Anything else, to me, is "a cherry on top".

**Comment by PeerHerholz** [comment]
mmmhhhh cherries... true that @ReinderVosDeWael! Based on that I would suggest to extract the needed information from the `atlas`/`nii` image itself. Currently I see three different "cases":

- no `parcellation`, but `voxel-wise`: use `NiftiMasker` to extract `voxel-wise` time series, compute `voxel` based `connectivity matrix`, compute `gradients`, use `NiftiMasker.inverse_transform` on `gradients` to map them back to `volume` space
- `probabilistic atlas` (4D): same as above but use `NiftiMapsMasker` and 4th dimension to map `vol2grad2vol`
 - `maximum probabilistic atlas` (3):  same as above but use `NiftiLabelsMasker` to map `vol2grad2vol`

**Comment by ReinderVosDeWael** [comment]
I think Peer covered all the relevant cases, or at least I don't see any others. Shall we go ahead with this? @PeerHerholz would you need help implementing this and if yes which parts do you need me to work on?

**Comment by llevitis** [comment]
hey @PeerHerholz and @ReinderVosDeWael...i'm v sorry for circling back to this so late, but I was wondering what the status is currently of the use cases that had been mentioned above?

**Comment by ReinderVosDeWael** [comment]
Thanks for reminding us of this :) I'll have a look after the OHBM abstract deadline
